### DISCUSSION TITLE
Optimize rhythmic drive

### DISCUSSION THREAD

**Discussion by samadpls** [jonescompneurolab/hnn-core#778] [Q&A]
Hi @carolinafernandezp,
I am @samadpls, a GSoC student this year working on a project. As an initial task for my mentorship, I have been assigned to understand and contribute to the ["Optimize rhythmic drives #673](https://github.com/jonescompneurolab/hnn-core/pull/673)" PR. After going through the PR description and the discussion, I have identified the following tasks that need to be completed:

The main issue that needs to be addressed is that the current implementation of the `_maximize_psd` objective function does not ensure that the resulting power spectral density (PSD) will express the specified ratios between the frequency bands. Even though the optimizer maximizes the power in the specified bands, it does not maintain the desired ratio between them, which is a crucial requirement.

1. Modify the `_maximize_psd` objective function
2. Implement the ability to pass a custom callable for `obj_fun`: The codebase currently allows users to pass `obj_fun_kwargs`, but not a custom callable function for `obj_fun`

3. Update documentation and tests and git conflicts

Please let me know if I'm on the right track with these tasks. I would appreciate your guidance and feedback on my understanding of the PR and the remaining work.

Additionally, I tried running the example code provided in the PR, but it is taking a long time to simulate. Are you aware of how long the simulation should typically take? If it's taking an unusually long time, I would appreciate any insights or suggestions you may have to troubleshoot the issue.

Thank you

**Comment by carolinafernandezp** [comment]
Hi,

1. The issue with `_maximize​_psd` not expressing the desired ratios was already addressed (I should have deleted that comment my apologies). You can see that it is resolved if you look in the PR description which has some sample code in which we wish to optimize power in the 9-11 Hz and 19-21 Hz bands in a ratio of 1 to 2 (i.e. twice as much power in the beta band as in the alpha band). Reversing the ratios also worked.
2. This issue was also already resolved. You can look at my comment from April 15th which mentions that the user can now pass a callable for `obj_fun`. I'm testing this in _test_general_optimization.py_, `test_user_obj_fun` where `obj_fun` is set to `maxime_csd` which is just a function that maximizes CSD in a certain time range and for a certain electrode depth range.
3. I'm currently waiting on other contributors to remove some GUI docs from the build (I believe) so that the build stops failing, but the documentation and tests are also done if you look under "Files changed."

In my opinion, the PR is done but it could be optimized after it is merged.

Another PR I think would be valuable to open/work on, would be one about optimizing Poisson drives (which I didn't get to do last summer). I would let the HNN team know and see where you can go from there.

Most of the code for this new PR would probably go under _objective_functions.py_. You would also need to come up with an example and add that under _examples/howto_, and finally add the relevant tests under _tests/test_general_optimization.py_.

Let me know if you ever want to schedule a call or feel free to reach out at any time!

Kindly

**Reply by samadpls** [reply]
Sure, I'd be happy to schedule a meeting to discuss the Poisson drive optimization PR once it's decided that the Poisson drive optimization needs to be worked on. I'm available anytime between 9 am and 4 pm EST, as I'm in the GMT+5 time zone. Additionally, I tried running the example provided in the PR, but it's taking too long to simulate. Are you aware of how long the simulation typically takes? Please let me know what time works best for you. Looking forward to our discussion!
